### PR TITLE
Wildcard query support

### PR DESCRIPTION
**Pull request by ashtul**
This PR adds support for wildcard queries for TEXT and TAG fields.
- `?` match any single character
- `*` match zero or more characters

The syntax is `w'foo?bar*baz'`.

The parser ignores all special characters but `'` and `\`(since there is a need to support escaping `'`). 

There is a new API introduced in `wildcard.h`
- `Wildcard_MatchChar`: Checks for a match between a pattern and a string. Returns -
  - `FULL_MATCH` - if a string matches a pattern.
  - `PARTIAL_MATCH` - if currently there is no match, but additional characters can lead to a match
  - `NO_MATCH` - if there is no match and additional character cannot lead to a match either.
- `Wildcard_MatchRune`: Same as Wildcard_MatchChar. Used by Trie instead of TrieMap.
- `Wildcard_TrimPattern`: Trim extra characters from the pattern. Ex. foo**bar -> foo*bar.
- `Wildcard_RemoveEscape`: Remove '\'. Required b/c parser support escaping `'`.
- `Wildcard_StarBreak`: To be used for support of SuffixTrie. Break pattern into token at `*`.

The logic in the iterators of `prefix`, `contains` and `suffix`, adds to the buffer one by one and checks for a potential match.
The `wildcard` iterator on both `Trie` and `TrieMap` adds all characters on the current node to the buffer and calls the` Wildcard_Match` function. The function evaluates the current buffer and determines whether a match was found or if additional characters can lead to a match. The return values are processed as follows:
- FULL_MATCH: a match was found, return it. If the pattern ends with `*`, return all the children as well.
- NO_MATCH: a match is not possible. Return.
- PARTIAL_MATCH: no match so far. If there are children, iterate on them.

Optimization:
- If a pattern does not include `*`, if the current buffer is the length of the pattern, the iteration on the branch is terminated.

MOD-2096

### CODE REVIEW

File: src/query.c [hunk 1]
```diff
@@ -202,6 +207,50 @@ QueryNode *NewPrefixNode_WithParams(QueryParseCtx *q, QueryToken *qt, bool prefi
...
+    // ret->verb.tok.len = _removeEscape(ret->verb.tok.str, ret->verb.tok.len);
+  }
+  return ret;
+}
+*/
+
+QueryNode *NewWildcardNode_WithParams(QueryParseCtx *q, QueryToken *qt) {
+  QueryNode *ret = NewQueryNode(QN_WILDCARD_QUERY);
+  q->numTokens++;
+  //printf("%s ", qt->s);
```
Comment (MeirShpilraien) [inline src/query.c:233]: Delete?

File: src/query.c [hunk 2]
```diff
@@ -202,6 +207,50 @@ QueryNode *NewPrefixNode_WithParams(QueryParseCtx *q, QueryToken *qt, bool prefi
...
+  if (qt->type == QT_WILDCARD) {
+    char *s = rm_malloc(qt->len + 1);
+    memcpy(s, qt->s, qt->len);
+    s[qt->len] = '\0';
+    // printf("%s ", s);
```
Comment (MeirShpilraien) [inline src/query.c:238]: Delete?

File: src/query.c [hunk 3]
```diff
@@ -202,6 +207,50 @@ QueryNode *NewPrefixNode_WithParams(QueryParseCtx *q, QueryToken *qt, bool prefi
...
+    //  size_t len = _removeEscape(s, qt->len);
+    // printf("%s", s);
+    ret->verb.tok = (RSToken){.str = s, .len = qt->len, .expanded = 0, .flags = 0};
+  } else {
+    assert(qt->type == QT_PARAM_WILDCARD);
+    QueryNode_InitParams(ret, 1);
+    QueryNode_SetParam(q, &ret->params[0], &ret->verb.tok.str, &ret->verb.tok.len, qt);
+    ret->params[0].type = PARAM_WILDCARD;
+    // qt->type = PARAM_WILDCARD;
```
Comment (MeirShpilraien) [inline src/query.c:247]: Delete?

File: src/query.c [hunk 4]
```diff
@@ -512,6 +560,53 @@ static IndexIterator *Query_EvalPrefixNode(QueryEvalCtx *q, QueryNode *qn) {
...
+
+  qn->verb.tok.len = Wildcard_RemoveEscape(qn->verb.tok.str, qn->verb.tok.len);
+  size_t nstr;
+  rune *str = strToFoldedRunes(qn->verb.tok.str, &nstr);
+
+  ctx.cap = 8;
+  ctx.its = rm_malloc(sizeof(*ctx.its) * ctx.cap);
+  ctx.nits = 0;
+
+  // spec support contains queries
```
Comment (MeirShpilraien) [inline src/query.c:586]: Delete?

File: src/query.c [hunk 5]
```diff
@@ -1081,6 +1237,10 @@ IndexIterator *Query_EvalNode(QueryEvalCtx *q, QueryNode *n) {
       return Query_EvalIdFilterNode(q, &n->fn);
     case QN_WILDCARD:
       return Query_EvalWildcardNode(q, n);
+    case QN_WILDCARD_QUERY:
+      return Query_EvalWildcardQueryNode(q,n);
+    case QN_VERBATIM:
+      //TODO:
```
Comment (MeirShpilraien) [inline src/query.c:1243]: todo?

File: src/query.c [hunk 6]
```diff
@@ -1201,6 +1360,9 @@ int QueryNode_EvalParams(dict *params, QueryNode *n, QueryError *status) {
   int withChildren = 1;
   int res = REDISMODULE_OK;
   switch(n->type) {
+    //case QN_WILDCARD_QUERY:
```
Comment (MeirShpilraien) [inline src/query.c:1363]: delete?

File: src/query.c [hunk 7]
```diff
@@ -1439,7 +1604,9 @@ static sds QueryNode_DumpSds(sds s, const IndexSpec *spec, const QueryNode *qs,
     case QN_FUZZY:
       s = sdscatprintf(s, "FUZZY{%s}\n", qs->fz.tok.str);
       return s;
-
+    case QN_WILDCARD_QUERY:
+    case QN_VERBATIM:
+      //TODO:
```
Comment (MeirShpilraien) [inline src/query.c:1609]: ?

File: src/query_node.h [hunk 8]
```diff
@@ -52,6 +52,12 @@ typedef enum {
   /* Vector */
   QN_VECTOR,
 
+  /* Verbatim */
+  QN_VERBATIM,
```
Comment (MeirShpilraien) [inline src/query_node.h:56]: Can you explain what this new QN is?

File: src/query_param.c [hunk 9]
```diff
@@ -136,6 +146,7 @@ int QueryParam_Resolve(Param *param, dict *params, QueryError *status) {
     return 0;
   size_t val_len;
   const char *val = Param_DictGet(params, param->name, &val_len, status);
+  // printf("val %s\n", val);
```
Comment (MeirShpilraien) [inline src/query_param.c:149]: delete

File: src/trie/rune_util.h [hunk 10]
```diff
@@ -63,6 +63,13 @@ static inline void runeBufFree(runeBuf *buf) {
   }
 }
 
+static inline void printfRune(const rune *rune, size_t len) {
```
Comment (MeirShpilraien) [inline src/trie/rune_util.h:66]: Do we use it somewhere? if not lets delete, if we do lets use RM_Log.

File: src/trie/trie.c [hunk 11]
```diff
@@ -1036,3 +1037,69 @@ static void containsIterate(TrieNode *n, t_len localOffset, t_len globalOffset,
...
+    r->stop = 1;
+  }
+  if (r->stop) {
+    return;
+  }
+
+  if (n->len != 0) { // not root
+    r->buf = array_ensure_append(r->buf, n->str, n->len, rune);
+  }
+  // printStats("append");
```
Comment (MeirShpilraien) [inline src/trie/trie.c:1053]: delete

File: src/trie/trie.c [hunk 12]
```diff
@@ -1036,3 +1037,69 @@ static void containsIterate(TrieNode *n, t_len localOffset, t_len globalOffset,
...
+      if (r->prefix) {
+        array_trimm_len(r->buf, n->len);
+        rangeIterateSubTree(n, r);
+        return; // we trimmed buffer earlier
+      } else {
+        // if node is terminal we add the result.
+        if (__trieNode_isTerminal(n)) {
+          r->callback(r->buf, array_len(r->buf), r->cbctx);
+        }
+        // we continue to look for matches on children similar to PARTIAL_MATCH
```
Comment (MeirShpilraien) [inline src/trie/trie.c:1071]: You mean we do not break right, maybe be specific about it.

File: deps/triemap/triemap.c [hunk 13]
```diff
@@ -1065,6 +1066,82 @@ int TrieMapIterator_NextContains(TrieMapIterator *it, char **ptr, tm_len_t *len,
...
+        case NO_MATCH: {
+          __tmi_Pop(it);
+          goto next;
+        }
+        case FULL_MATCH: {
+          // if query string ends with *, all following children are a match
+          if (it->buf[array_len(it->buf) - 1] == '*') {
+            current->found = true;
+          }
+          // current node is a term and should be returned
```
Comment (MeirShpilraien) [inline deps/triemap/triemap.c:1107]:
```suggestion
          // current node is a terminal and should be returned
```
I guess this is what you meant? Or I got it wrong...

File: deps/triemap/triemap.c [hunk 14]
```diff
@@ -1065,6 +1066,82 @@ int TrieMapIterator_NextContains(TrieMapIterator *it, char **ptr, tm_len_t *len,
   return 0;
 }
 
+int TrieMapIterator_NextWildcard(TrieMapIterator *it, char **ptr, tm_len_t *len, void **value) {
+  if (TimedOut_WithCounter(&it->timeout, &it->timeoutCounter)) {
+    return 0;
+  }
+
+  while (array_len(it->stack) > 0) {
```
Comment (MeirShpilraien) [inline deps/triemap/triemap.c:1074]: Shouldn't we check for timeout inside the loop?

File: deps/triemap/triemap.h [hunk 15]
```diff
@@ -164,8 +166,12 @@ void TrieMapIterator_Free(TrieMapIterator *it);
 int TrieMapIterator_Next(TrieMapIterator *it, char **ptr, tm_len_t *len, void **value);
 
 /* Iterate to the next matching entry in the trie. Returns 1 if we can continue,
- * or 0 if we're done and should exit */
+ * or 0 if we're done and should exit 
+ * NextContains is used by Contains and Suffix queries.
+ * Wildcard is used by Wildcard queries.
+ */
 int TrieMapIterator_NextContains(TrieMapIterator *it, char **ptr, tm_len_t *len, void **value);
+int TrieMapIterator_NextWildcard(TrieMapIterator *it, char **ptr, tm_len_t *len, void **value);
```
Comment (MeirShpilraien) [inline deps/triemap/triemap.h:174]: How will you know that a timeout happened? You will not get it from the return value ...

File: deps/wildcard/wildcard.c [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+  while (++i) {
+    //printf("%d ", i);
+    if (pattern_end != pattern_itr) {
+      const char c = *pattern_itr;
+      if ((str_end != str_itr) && (c == *str_itr || c == '?')) {
+        ++str_itr;
+        ++pattern_itr;
+        continue;
+      } else if (c == '*') {
+        while ((pattern_end != pattern_itr) && (*pattern_itr == '*')) {
```
Comment (MeirShpilraien) [inline deps/wildcard/wildcard.c:24]: `pattern_end` points to one char after the `pattern` buffer, so here you might access invalid memory (unless you assume its a C terminated strings and you have the `\0` ah the end, in such case why you pass the `len`).
General note, I believe that Redis already has a pattern matching function that is used on `keys` and `scan` command, why don't we take it?

Comment (MeirShpilraien) [reply deps/wildcard/wildcard.c:24]: Also took that code to gears:
https://github.com/RedisGears/RedisGears/blob/7927cd41d560a810a54dcaf7170cc172355a4a15/src/readers/keys_reader.c#L558

File: deps/wildcard/wildcard.c [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+size_t Wildcard_TrimPattern(char *pattern, size_t p_len) {
+  size_t i = 0;
+  size_t runner = 0;
+
+  while (i < p_len) {
+    if (pattern[i] == '*') {
+      // skip following starts
+      while (pattern[i + 1] == '*') {
```
Comment (MeirShpilraien) [inline deps/wildcard/wildcard.c:103]: What if `i` points to the last char, again you assume C string here?

File: deps/wildcard/wildcard.c [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+      return NO_MATCH;
+    }
+    pattern_itr = np_itr;
+    str_itr = ns_itr;
+  }
+  assert(0);
+  return FULL_MATCH;
+}
+
+size_t Wildcard_TrimPattern(char *pattern, size_t p_len) {
```
Comment (MeirShpilraien) [inline deps/wildcard/wildcard.c:96]: Can you add some documentation about how the trimming is work and why the it is fine to apply the logic (why the pattern before and after the logic are equivalent...)

File: deps/wildcard/wildcard.c [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+    if (str[runner] == '\0') {
+      break;
+    }
+  }
+
+  str[runner] = '\0';
+  return runner;
+}
+
+int Wildcard_StarBreak(const char *str, size_t len, size_t *tokenIdx, size_t *tokenLen) {
```
Comment (MeirShpilraien) [inline deps/wildcard/wildcard.c:148]: Can you explain this function and why you need it?

File: src/query.c [hunk 20]
```diff
@@ -916,6 +1011,64 @@ static IndexIterator *Query_EvalTagPrefixNode(QueryEvalCtx *q, TagIndex *idx, Qu
...
+
+  // an upper limit on the number of expansions is enforced to avoid stuff like "*"
+  char *s;
+  tm_len_t sl;
+  void *ptr;
+
+  // Find all completions of the prefix
+  while (TrieMapIterator_NextWildcard(it, &s, &sl, &ptr) &&
+        (itsSz < RSGlobalConfig.maxPrefixExpansions)) {
+          // TODO: use NewIndexReaderGeneric
```
Comment (MeirShpilraien) [inline src/query.c:1042]: Is this todo still relevant?

File: src/query.c [hunk 21]
```diff
@@ -916,6 +1011,64 @@ static IndexIterator *Query_EvalTagPrefixNode(QueryEvalCtx *q, TagIndex *idx, Qu
...
+    //NewIndexReaderGeneric(ptr, q->sctx->spec, )
+    IndexIterator *ret = TagIndex_OpenReader(idx, q->sctx->spec, s, sl, 1);
+    if (!ret) continue;
+
+    // Add the reader to the iterator array
+    its[itsSz++] = ret;
```
Comment (MeirShpilraien) [inline src/query.c:1048]: Can we use the arr.h here?

File: src/query_node.h (hunk 8 above)
Comment (ashtul) [reply src/query_node.h:56]: Removed

I placed it when I thought we would have `'fooo'` as a token w/o need to escape special chars (but `'` and `\`)

File: src/trie/rune_util.h (hunk 10 above)
Comment (ashtul) [reply src/trie/rune_util.h:66]: it is nice to have for debug since it is hard to read `rune` itself.

File: deps/triemap/triemap.c (hunk 14 above)
Comment (ashtul) [reply deps/triemap/triemap.c:1074]: Good point.

File: deps/triemap/triemap.h (hunk 15 above)
Comment (ashtul) [reply deps/triemap/triemap.h:174]: That is true for all timeouts we may suffer during the build of the iteration tree.
We do check for timeout once `QAST_Iterate` call is done and if true, set status to `QUERY_TIMEDOUT`.

File: deps/wildcard/wildcard.c (hunk 16 above)
Comment (ashtul) [reply deps/wildcard/wildcard.c:24]: I will look into the other functions.

File: deps/wildcard/wildcard.c (hunk 17 above)
Comment (ashtul) [reply deps/wildcard/wildcard.c:103]: Yes. We assume C string

File: deps/wildcard/wildcard.c (hunk 19 above)
Comment (ashtul) [reply deps/wildcard/wildcard.c:148]: Explanations are on the header file

File: src/query.c [hunk 22]
```diff
@@ -916,6 +1011,64 @@ static IndexIterator *Query_EvalTagPrefixNode(QueryEvalCtx *q, TagIndex *idx, Qu
...
+  }
+  TrieMapIterator_Free(it);
+
+  // printf("Expanded %d terms!\n", itsSz);
+  if (itsSz == 0) {
+    rm_free(its);
+    return NULL;
+  }
+  if (itsSz == 1) {
+    // TODO:
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/query.c:1062]: This was copy-pasted from above.
It is about getting rid of the union iterator and just placing the iterator itself. I didn't do it in the past b/c of the profile iterator.
WDYT?

File: src/query.c (hunk 21 above)
Comment (ashtul) [reply src/query.c:1048]: We use the same logic on all iterators above. IMO, no need to change here.

File: deps/wildcard/wildcard.c (hunk 16 above)
Comment (ashtul) [reply deps/wildcard/wildcard.c:24]: redis is using the `glob` library. This is a good solution for `char` strings but not for `rune` strings.

File: deps/wildcard/wildcard.c (hunk 17 above)
Comment (MeirShpilraien) [reply deps/wildcard/wildcard.c:103]: So why we need the len? We can just check for `\0` to know if we got to the end no?

File: src/trie/rune_util.h (hunk 10 above)
Comment (MeirShpilraien) [reply src/trie/rune_util.h:70]: Do you think we might need it to debug production issues? If so lets make a debug command out of it?

File: deps/wildcard/wildcard.c (hunk 17 above)
Comment (ashtul) [reply deps/wildcard/wildcard.c:103]: Left comment

File: src/trie/rune_util.h (hunk 10 above)
Comment (ashtul) [reply src/trie/rune_util.h:70]: This is internal. The print out to the cli will be identical to the original. It is more useful when we scan it and try to understand what is the current state.

**Review by MeirShpilraien — DISMISSED** [review]

**Review by MeirShpilraien — APPROVED** [review]

**Pull request merged**